File: 17-Letter Combinations of a Phone Number-电话号码的字母组合/17-Letter Combinations of a Phone Number-电话号码的字母组合.java
class Solution {
    private List<String> res;
    private StringBuilder path = new StringBuilder();
    String[] table = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    public List<String> letterCombinations(String digits) {
        res = new ArrayList<String>();
        if(digits.length() == 0) return res;
        backtrace(digits, 0);
        return res;
    }

    public void backtrace(String digits, int num){
        // 每个递归代表digits中的一个数
        if(path.length() == digits.length()) {
        	res.add(path.toString());
            return;
        }
        
        // 每次循环代表遍历该数对应的码表中的每个数
        String curString = table[digits.charAt(num) - '0'];
        for(int i = 0; i < curString.length(); i++) {
        	path.append(curString.charAt(i));
        	backtrace(digits, num+1);
        	path.deleteCharAt(path.length() - 1);
        }
    }
}

File: 17-Letter Combinations of a Phone Number-电话号码的字母组合/17-Letter Combinations of a Phone Number-电话号码的字母组合.js
/**
 * @param {string} digits
 * @return {string[]}
 */
 let res = []
 let path = []
 let reflectTable = {
    '0':[],
    '1': [],
    '2': ['a', 'b', 'c'],
    '3': ['d', 'e', 'f'],
    '4': ['g', 'h', 'i'],
    '5': ['j', 'k', 'l'],
    '6': ['m', 'n', 'o'],
    '7': ['p', 'q', 'r', 's'],
    '8': ['t', 'u', 'v'],
    '9': ['w', 'x', 'y', 'z'],
 }
var letterCombinations = function(digits) {
    res = []
    if(digits.length === 0) return res
    backtrace(digits, 0)
    return res
};

var backtrace = function(digits, index){
    if(path.length === digits.length){
        res.push(path.join(""))
        return
    }

    let curArr = reflectTable[digits[index]]
    for(let i = 0; i < curArr.length; i++){
        path.push(curArr[i])
        backtrace(digits, index + 1)
        path.pop()
    }
}